libyul/backends/wasm/EWasmCodeTransform.cpp

```diff
@@ -0,0 +1,317 @@
+		return move(assignment);
+
+	wasm::Block block;
+	block.statements.emplace_back(std::move(assignment));
+	for (size_t i = 1; i < _variableNames.size(); ++i)
+		block.statements.emplace_back(wasm::LocalAssignment{
+			std::move(_variableNames.at(i)),
+			make_unique<wasm::Expression>(wasm::GlobalVariable{m_globalVariables.at(i - 1).variableName})
+		});
+	return std::move(block);
+}
+
+wasm::Expression EWasmCodeTransform::operator()(VariableDeclaration const& _varDecl)
+{
+	vector<string> variableNames;
+	for (auto const& var: _varDecl.variables)
+	{
+		variableNames.emplace_back(var.name.str());
+		m_localVariables.emplace_back(wasm::VariableDeclaration{variableNames.back()});
+	}
+
+	if (_varDecl.value)
+		return generateMultiAssignment(std::move(variableNames), visit(*_varDecl.value));
+	else
+		return wasm::BuiltinCall{"nop", {}};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Assignment const& _assignment)
+{
+	vector<string> variableNames;
+	for (auto const& var: _assignment.variableNames)
+		variableNames.emplace_back(var.name.str());
+	return generateMultiAssignment(std::move(variableNames), visit(*_assignment.value));
+}
+
+wasm::Expression EWasmCodeTransform::operator()(StackAssignment const&)
+{
+	yulAssert(false, "");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(ExpressionStatement const& _statement)
+{
+	return visitReturnByValue(_statement.expression);
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Label const&)
+{
+	yulAssert(false, "");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionalInstruction const&)
+{
+	yulAssert(false, "");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionCall const& _call)
+{
+	if (m_dialect.builtin(_call.functionName.name))
+		return wasm::BuiltinCall{_call.functionName.name.str(), visit(_call.arguments)};
+	else
+		// If this function returns multiple values, then the first one will
+		// be returned in the expression itself and the others in global variables.
+		// The values have to be used right away in an assignment or variable declaration,
+		// so it is handled there.
+		return wasm::FunctionCall{_call.functionName.name.str(), visit(_call.arguments)};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Identifier const& _identifier)
+{
+	return wasm::LocalVariable{_identifier.name.str()};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Literal const& _literal)
+{
+	u256 value = valueOfLiteral(_literal);
+	yulAssert(value <= numeric_limits<uint64_t>::max(), "");
+	return wasm::Literal{uint64_t(value)};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(yul::Instruction const&)
+{
+	yulAssert(false, "EVM instruction used for Wasm code generation.");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(If const& _if)
+{
+	return wasm::If{visit(*_if.condition), visit(_if.body.statements), {}};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Switch const& _switch)
```

Comment:
FWIW, `Switch` statements probably will be transformed into `If` statements in earlier phases.